com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/graph/ResourceNode.java

```diff
@@ -0,0 +1,150 @@
+
+    public ResourceNode(IResource resource) {
+        this(resource.getPath(), resource.output().getAbsPath());
+        this.resource = resource;
+
+    }
+    public ResourceNode(final String relativeFilepath, final String absoluteFilepath) {
+        if (relativeFilepath.startsWith("/")) {
+            this.relativeFilepath = relativeFilepath;
+        } else {
+            this.relativeFilepath = "/" + relativeFilepath;
+        }
+        this.absoluteFilepath = absoluteFilepath;
+    }
+
+    public void addChild(ResourceNode childNode) {
+        childNode.parent = this;
+        this.children.add(childNode);
+    }
+
+    public void addChildren(List<ResourceNode> childNodes) {
+        for (ResourceNode childNode : childNodes) {
+            childNode.parent = this;
+            this.children.add(childNode);
+        }
+    }
+
+    public List<ResourceNode> getChildren() {
+        return this.children;
+    }
+
+    public ResourceNode getParent() {
+        return this.parent;
+    }
+
+    public IResource getResource() {
+        return this.resource;
+    }
+
+    public String getRelativePath() {
+        return relativeFilepath;
+    }
+
+    public String getAbsolutePath() {
+        return absoluteFilepath;
+    }
+
+    public void setHexDigest(String hexDigest) {
+        this.hexDigest = hexDigest;
+    }
+
+    public void setExcluded(boolean excluded) {
+        this.excluded = excluded;
+    }
+
+    private void writeJSON(JsonGenerator generator) throws IOException {
+        generator.writeStartObject();
+        generator.writeFieldName("path");
+        generator.writeString(relativeFilepath);
+
+        generator.writeFieldName("hexDigest");
+        generator.writeString(hexDigest);
+
+        generator.writeFieldName("excluded");
+        generator.writeBoolean(excluded);
+
+        generator.writeFieldName("children");
+        generator.writeStartArray();
+        for (ResourceNode child : children) {
+            child.writeJSON(generator);
+        }
+        generator.writeEndArray();
+        generator.writeEndObject();
+    }
+
+    public String toJSON() throws IOException {
+        StringWriter stringWriter = new StringWriter();
+        BufferedWriter writer = new BufferedWriter(stringWriter);
+        JsonGenerator generator = null;
+        try {
+            generator = (new JsonFactory()).createJsonGenerator(writer);
+            generator.useDefaultPrettyPrinter();
+            writeJSON(generator);
+        }
+        finally {
+            if (generator != null) {
+                generator.close();
+            }
+            IOUtils.closeQuietly(writer);
+        }
+        return stringWriter.toString();
+    }
+
+    @Override
+    public String toString() {
+        return "[ResourceNode path: " + relativeFilepath + "]";
+    }
+
+    @Override
+    public int hashCode() {
+        return absoluteFilepath.hashCode();
```

Comment:
I'm not sure why we rely on absolute paths?
Relative should be fine (they're still unique).

Shorter paths are shorter (good),

But more importantly, relative paths are generally more useful when working with built content, as you may want to move resources (locally, or use a path over a remote network etc)